File: typescript/src/seams.ts
/**
 * The four seams, and the value model they pass between them.
 *
 * Each argument of the two calls is one job, so a format can be added without
 * inventing a source and a source without inventing a format (FR-2.5).
 *
 *   Codec   decode(bytes) -> value, encode(value) -> bytes
 *           The format. Knows nothing about where bytes came from.
 *           `encode` emits canonical form.
 *   Reader  read(path) -> bytes
 *           The IO on the way in. HANDED THE PATH, NOT THE BYTES.
 *   Writer  write(path, bytes)
 *           The IO on the way out. Puts the whole contents in place.
 *   Schema  validate(value)
 *           Applies to the decoded structure, not the text.
 *
 * A READER IS HANDED THE PATH RATHER THAN BYTES (FR-2.5a). Handing it bytes
 * would put the open where the caller is, and a test substituting a reader
 * would then be replacing the parse alone. Handed the path, a substituted
 * reader exercises every validation path against no filesystem at all.
 *
 * They are declared as interfaces because TypeScript has them and a consumer
 * gets the check for free. Nothing here asks for a class: any object with the
 * method is the seam, which is what lets a test supply four lines instead of an
 * implementation.
 */

/**
 * What a decoder produces and an encoder accepts, and nothing else (FR-2.9).
 *
 * Maps, lists and the JSON scalars. A format with a type JSON does not have is
 * reconciled in the decoder: a lossless spelling is coerced, and a mapping key
 * that is not a string is refused rather than stringified, because that
 * coercion is not reversible.
 *
 * JavaScript's `number` is the JSON number, so this pack needs no rule for
 * telling an integer from a float in the value model. What that costs is stated
 * in `float_text.ts`.
 */
export type WrenchValue =
  | null
  | boolean
  | number
  | string
  | WrenchValue[]
  | { [key: string]: WrenchValue };

/** The format. Bytes to a structure, and a structure to canonical bytes. */
export interface Codec {
  /**
   * Bytes into maps, lists and scalars.
   *
   * Text is accepted as well as bytes, because a substituted reader in a test
   * has no filesystem to have read bytes from and writing a string is what
   * makes that test four lines.
   */
  decode(data: Uint8Array | string): WrenchValue;
  /** A structure into canonical bytes. */
  encode(value: WrenchValue): Uint8Array;
}

/** The IO on the way in, handed the path rather than the bytes. */
export interface Reader {
  read(path: string): Uint8Array | string;
}

/** The IO on the way out, putting the whole contents in place. */
export interface Writer {
  write(path: string, bytes: Uint8Array): void;
}

/** Applied to the decoded structure rather than to the text. */
export interface Schema {
  validate(value: WrenchValue): void;
}

/** Bytes, whichever of the two a seam supplied. */
export function asText(data: Uint8Array | string): string {
  return typeof data === "string" ? data : new TextDecoder().decode(data);
}

File: ruby/lib/wrench/seams.rb
# frozen_string_literal: true

require "fileutils"
require "tmpdir"

require_relative "errors"

module Wrench
  # The four seams. Each argument of the two calls is one job, so a format can
  # be added without inventing a source and a source without inventing a format.
  #
  # Ruby needs no declared interface for this and one is not invented: a codec
  # is anything answering `decode` and `encode`, a reader anything answering
  # `read`. What is written down instead is the contract each promises, because
  # that is the part another pack has to agree with.
  #
  #   Codec   decode(bytes) -> value, encode(value) -> bytes
  #           The format. Knows nothing about where bytes came from.
  #           `encode` emits canonical form.
  #   Reader  read(path) -> bytes
  #           The IO on the way in. HANDED THE PATH, NOT THE BYTES.
  #   Writer  write(path, bytes)
  #           The IO on the way out. Puts the whole contents in place.
  #   Schema  validate(value)
  #           Applies to the decoded structure, not the text.
  #
  # A READER IS HANDED THE PATH RATHER THAN BYTES, FR-2.5a. Handing it bytes
  # would put the open where the caller is, and a test substituting a reader
  # would then be replacing the parse alone. Handed the path, a substituted
  # reader exercises every validation path against no filesystem at all.

  # The one reader and writer that ship, both local files, and nothing else
  # (FR-2.8). Everything wrench serves runs on the machine holding the file, and
  # a test brings its own.
  class LocalFileIO
    # The whole contents as bytes.
    def read(path)
      File.binread(path.to_s)
    rescue SystemCallError, IOError => e
      raise ReadError.new("could not read the file", cause: e, path: path)
    end

    # The bytes, in place, atomically (FR-6.3).
    #
    # They go to a temporary file BESIDE the target and are renamed over it, so
    # a concurrent reader sees the previous contents or the new ones and never a
    # half-written file. Beside rather than in a temporary directory, because a
    # rename across filesystems is a copy and stops being atomic.
    def write(path, bytes)
      target = File.expand_path(path.to_s)
      directory = File.dirname(target)
      temporary = nil
      begin
        File.open(File.join(directory, ".wrench-#{Process.pid}-#{object_id}.tmp"), "wb") do |f|
          temporary = f.path
          f.write(bytes)
          f.flush
          f.fsync
        end
        File.rename(temporary, target)
        temporary = nil
      rescue SystemCallError, IOError => e
        raise WriteError.new("could not write the file", cause: e, path: path)
      ensure
        FileUtils.rm_f(temporary) if temporary
      end
    end
  end

  # The shipped reader and writer, as one object, which is what a caller passes
  # for both seams.
  LOCAL_FILE = LocalFileIO.new
end
